File: uranium-arkade-canvas/dom-api/src/wasm32Main/js/nativeCanvas.js
konan.libraries.push({
    uranium_getNativeCanvas: (canvasIdPtr, canvasIdLength, resultArena) => {
        const canvasId = toUTF16String(canvasIdPtr, canvasIdLength);
        const canvas = document.getElementById(canvasId);
        if(!canvas) throw Error('Canvas ' + canvasId + ' not found')
        return newArenaObject(resultArena, canvas);
    },
    uranium_NativeCanvas_getContext: (canvasArena, canvasIndex, resultArena) => {
        const canvas = kotlinObject(canvasArena, canvasIndex);
        return newArenaObject(resultArena, canvas.getContext('2d'));
    },
    uranium_NativeCanvas_getWidth: (canvasArena, canvasIndex) =>
        kotlinObject(canvasArena, canvasIndex).width,
    uranium_NativeCanvas_setWidth: (canvasArena, canvasIndex, width) =>
        kotlinObject(canvasArena, canvasIndex).width = width,
    uranium_NativeCanvas_getHeight: (canvasArena, canvasIndex) =>
        kotlinObject(canvasArena, canvasIndex).height,
    uranium_NativeCanvas_setHeight: (canvasArena, canvasIndex, height) =>
        kotlinObject(canvasArena, canvasIndex).height = height,
    uranium_NativeCanvas_getClientWidth: (canvasArena, canvasIndex) =>
        kotlinObject(canvasArena, canvasIndex).clientWidth,
    uranium_NativeCanvas_getClientHeight: (canvasArena, canvasIndex) =>
        kotlinObject(canvasArena, canvasIndex).clientHeight,
    uranium_NativeCanvas_setOnMouseDownListener: (canvasArena, canvasIndex, handlerIndex, handlerArena) =>
        kotlinObject(canvasArena, canvasIndex).onmousedown = wrapLambda(handlerArena, handlerIndex),
    uranium_NativeCanvas_setOnMouseMoveListener: (canvasArena, canvasIndex, handlerIndex, handlerArena) =>
        kotlinObject(canvasArena, canvasIndex).onmousemove = wrapLambda(handlerArena, handlerIndex),
    uranium_NativeCanvas_setOnMouseUpListener: (canvasArena, canvasIndex, handlerIndex, handlerArena) =>
        kotlinObject(canvasArena, canvasIndex).onmouseup = wrapLambda(handlerArena, handlerIndex),
    uranium_NativeCanvas_setOnMouseEnterListener: (canvasArena, canvasIndex, handlerIndex, handlerArena) =>
        kotlinObject(canvasArena, canvasIndex).onmouseenter = wrapLambda(handlerArena, handlerIndex),
    uranium_NativeCanvas_setOnMouseLeaveListener: (canvasArena, canvasIndex, handlerIndex, handlerArena) =>
        kotlinObject(canvasArena, canvasIndex).onmouseleave = wrapLambda(handlerArena, handlerIndex),
});

File: uranium-arkade-htmlcanvas/dom-api/src/wasm32Main/kotlin/pl/karol202/uranium/arkade/htmlcanvas/dom/canvas/NativeCanvas.kt
package pl.karol202.uranium.arkade.htmlcanvas.dom.canvas

import kotlinx.wasm.jsinterop.*
import pl.karol202.uranium.arkade.htmlcanvas.dom.NativeValue
import pl.karol202.uranium.arkade.htmlcanvas.dom.input.NativeMouseEvent
import pl.karol202.uranium.arkade.htmlcanvas.dom.retrieveObject

@SymbolName("uranium_getNativeCanvas")
private external fun getNativeCanvas(canvasIdPtr: Pointer, canvasIdLength: Int, resultArena: Arena): Int

@SymbolName("uranium_NativeCanvas_getContext")
private external fun NativeCanvas_getContext(canvasArena: Arena, canvasIndex: Int, resultArena: Arena): Int

@SymbolName("uranium_NativeCanvas_getWidth")
private external fun NativeCanvas_getWidth(canvasArena: Arena, canvasIndex: Object): Int

@SymbolName("uranium_NativeCanvas_setWidth")
private external fun NativeCanvas_setWidth(canvasArena: Arena, canvasIndex: Object, width: Int)

@SymbolName("uranium_NativeCanvas_getHeight")
private external fun NativeCanvas_getHeight(canvasArena: Arena, canvasIndex: Object): Int

@SymbolName("uranium_NativeCanvas_setHeight")
private external fun NativeCanvas_setHeight(canvasArena: Arena, canvasIndex: Object, height: Int)

@SymbolName("uranium_NativeCanvas_getClientWidth")
private external fun NativeCanvas_getClientWidth(canvasArena: Arena, canvasIndex: Object): Int

@SymbolName("uranium_NativeCanvas_getClientHeight")
private external fun NativeCanvas_getClientHeight(canvasArena: Arena, canvasIndex: Object): Int

@SymbolName("uranium_NativeCanvas_setOnMouseDownListener")
private external fun NativeCanvas_setOnMouseDownListener(canvasArena: Arena, canvasIndex: Object,
                                                         handlerIndex: Int, handlerArena: Arena)

@SymbolName("uranium_NativeCanvas_setOnMouseMoveListener")
private external fun NativeCanvas_setOnMouseMoveListener(canvasArena: Arena, canvasIndex: Object,
                                                         handlerIndex: Int, handlerArena: Arena)

@SymbolName("uranium_NativeCanvas_setOnMouseUpListener")
private external fun NativeCanvas_setOnMouseUpListener(canvasArena: Arena, canvasIndex: Object,
                                                       handlerIndex: Int, handlerArena: Arena)

@SymbolName("uranium_NativeCanvas_setOnMouseEnterListener")
private external fun NativeCanvas_setOnMouseEnterListener(canvasArena: Arena, canvasIndex: Object,
                                                          handlerIndex: Int, handlerArena: Arena)

@SymbolName("uranium_NativeCanvas_setOnMouseLeaveListener")
private external fun NativeCanvas_setOnMouseLeaveListener(canvasArena: Arena, canvasIndex: Object,
                                                          handlerIndex: Int, handlerArena: Arena)

actual class NativeCanvas(override val arena: Arena,
                          override val index: Object) : NativeValue
{
	actual val context: NativeCanvasContext
		get() = retrieveObject(::NativeCanvasContext) {
			NativeCanvas_getContext(arena, index, resultArena)
		}

	actual var width
		get() = NativeCanvas_getWidth(arena, index)
		set(value) { NativeCanvas_setWidth(arena, index, value) }
	actual var height
		get() = NativeCanvas_getHeight(arena, index)
		set(value) { NativeCanvas_setHeight(arena, index, value) }
	actual val clientWidth get() = NativeCanvas_getClientWidth(arena, index)
	actual val clientHeight get() = NativeCanvas_getClientHeight(arena, index)

	actual fun setOnMouseDownListener(listener: ((NativeMouseEvent) -> Unit)?) =
			NativeCanvas_setOnMouseDownListener(arena, index, wrapFunction(mouseHandler(listener)), ArenaManager.globalArena)

	actual fun setOnMouseMoveListener(listener: ((NativeMouseEvent) -> Unit)?) =
			NativeCanvas_setOnMouseMoveListener(arena, index, wrapFunction(mouseHandler(listener)), ArenaManager.globalArena)

	actual fun setOnMouseUpListener(listener: ((NativeMouseEvent) -> Unit)?) =
			NativeCanvas_setOnMouseUpListener(arena, index, wrapFunction(mouseHandler(listener)), ArenaManager.globalArena)

	actual fun setOnMouseEnterListener(listener: ((NativeMouseEvent) -> Unit)?) =
			NativeCanvas_setOnMouseEnterListener(arena, index, wrapFunction(mouseHandler(listener)), ArenaManager.globalArena)

	actual fun setOnMouseLeaveListener(listener: ((NativeMouseEvent) -> Unit)?) =
			NativeCanvas_setOnMouseLeaveListener(arena, index, wrapFunction(mouseHandler(listener)), ArenaManager.globalArena)

	private fun mouseHandler(listener: ((NativeMouseEvent) -> Unit)?): KtFunction<Unit> = { (eventJsValue) ->
		listener?.invoke(NativeMouseEvent(eventJsValue.arena, eventJsValue.index))
	}
}

actual fun getNativeCanvas(canvasId: String) = retrieveObject(::NativeCanvas) {
	getNativeCanvas(stringPointer(canvasId), stringLengthBytes(canvasId), resultArena)
}
